File: code/Tree.cpp
#include <iostream>
#include <string>
#include <bits/stdc++.h>
#include <iostream>
#include <string>
#include <cstdio>
#include <bits/stdc++.h>
#include <algorithm>
#include <typeinfo>
#include <utility>
#include <vector>
#include <unordered_set>
using namespace std;
class BinaryTree{
    public:
    int data;
    BinaryTree* left;
    BinaryTree* right;

    BinaryTree(int data){
        this->data=data;
        this->left=NULL;
        this->right=NULL;
    }
};

BinaryTree* Tree(BinaryTree* root){
    int data;
    
    cout<<"enter data: "<<endl;
    cin>>data;
    if(data==-1){
        return NULL;
    }
    root=new BinaryTree(data);

    cout<<"enter data for inserting in left of: "<<data<<endl;
    root->left=Tree(root->left);
    cout<<"enter data for inserting in right of: "<<data<<endl;
    root->right=Tree(root->right);
    return root;
}

void level_order_traversal(BinaryTree* root){
    queue<BinaryTree*> q;
    q.push(root);
    q.push(NULL);//separator

    while(!q.empty()){
        BinaryTree* temp=q.front();
        
        q.pop();

        if(temp==NULL){
            cout<<endl;
            if(!q.empty()){
                q.push(NULL);
            }
        }
        else{
            cout<<temp->data<<" ";
            if(temp->left){
                q.push(temp->left);
            }
            if(temp->right){
                q.push(temp->right);
            }
        }  
    }
}

//left root right
void inorder_traversal(BinaryTree* root){
    BinaryTree* temp=root;
    if(temp->left==NULL){
        cout<<temp->data<<" ";
        return ;
    }
    if(temp->right==NULL){
        cout<<temp->data<<" ";
        return ;
    }
    inorder_traversal(temp->left);
    cout<<temp->data<<" ";
    inorder_traversal(temp->right);
}

//left right root
void post_order_traversal(BinaryTree* root){
    BinaryTree* temp=root;
    if(temp->left==NULL){
        cout<<temp->data<<" ";
        return ;
    }
    if(temp->right==NULL){
        cout<<temp->data<<" ";
        return ;
    }
    post_order_traversal(temp->left);
    
    post_order_traversal(temp->right);
    cout<<temp->data<<" ";
}

//root left right 
void pre_order_traversal(BinaryTree* root){
    BinaryTree* temp=root;
    if(temp->left==NULL){
        cout<<temp->data<<" ";
        return ;
    }
    if(temp->right==NULL){
        cout<<temp->data<<" ";
        return ;
    }
    cout<<temp->data<<" ";
    pre_order_traversal(temp->left);
    
    pre_order_traversal(temp->right);
}

int height(BinaryTree* root){
    if(root==NULL){
        return 0;
    }

    int l=height(root->left);
    int r=height(root->right);
    int ans =max(l,r)+1;
    return ans;
}

vector<int> spiral_traversal(BinaryTree* root){
    bool ori=true;//left to right
    vector<int> ans;
    if (root==NULL){
        return ans;      
    }
    queue<BinaryTree*> q;
    q.push(root);
    
    while(!q.empty()){
        int size=q.size();
        vector<int> temp(size);
        for(int i=0;i<size;i++){
            BinaryTree* Front=q.front();
            q.pop();

            int index=ori?i:size-i-1;
            temp[index]=Front->data;

            if(Front->left){
                q.push(Front->left);
            }
            if(Front->right){
                q.push(Front->right);
            }
        }
        ori=!ori;
        for(auto i:temp){
            ans.push_back(i);
        }
    }
    return ans;
}
vector<int> verticalTraversal(BinaryTree* node){
    map<int,map<int,vector<int>>> nodes;
    queue<pair<BinaryTree*,pair<int,int>>> q;
    vector<int> ans;
    if(node==NULL){
        return ans;
    }

    q.push(make_pair(node,make_pair(0,0)));
    while(!q.empty()){
        pair<BinaryTree*,pair<int,int>> temp=q.front();
        q.pop();
        BinaryTree* front_node=temp.first;
        int hd=temp.second.first;
        int lvl=temp.second.second;
        nodes[hd][lvl].push_back(front_node->data);
        if(front_node->left){
            q.push(make_pair(front_node->left,make_pair(hd-1,lvl+1)));
        }
        if(front_node->right){
            q.push(make_pair(front_node->right,make_pair(hd+1,lvl+1)));
        }
    }
    cout<<"map";
    for(auto i:nodes){
        cout<<i.first<<" ";
        for(auto j:i.second){
            cout<<j.first<<" ";
            for(auto z:j.second){
                cout<<z<<" ";
            }
            // for(auto k:j.second){
                
            //     ans.push_back(k);
            // }
        }
        cout<<endl;
    }
    
    
    return ans;
    

}

bool identical(BinaryTree* root1,BinaryTree* root2 ){
    
    if(root1==nullptr && root2==nullptr){
        return true;
    }
    if(root1==nullptr || root2==nullptr){
        return false;
    }

   return (root1->data == root2->data) && identical(root1->left,root2->left) && identical(root1->right,root2->right);
}

int height_depth(BinaryTree* root){
        if (root == nullptr){
            return 0;
        }
        
        int left=height_depth(root->left);
        int right = height_depth(root->right);
        int maxy = max(left,right)+1;
        return maxy;
    }

int diameter(BinaryTree* root,int maxi){ //time complexity O(n**2)
    if(root==nullptr){
        return 0;
    }
    int lh = height_depth(root->left);
    int rh = height_depth(root->right);
    maxi = max(maxi,lh+rh+1);
    diameter(root->left,maxi);
    diameter(root->right,maxi);
    return maxi;

}

int diameter_2(BinaryTree* root,int maxi){ // optimized version O(n)
    if(root==nullptr){
        return 0;
    }
    int left = diameter_2(root->left,maxi);
    int right = diameter_2(root->right,maxi);
    maxi = max(maxi, left+right);
    return 1+max(left,right);
}

int maximum_sum_path(BinaryTree* root, int &maxi){ // O(n)
    if(root==nullptr){
        return 0;
    }
    int left_sum = max(0,maximum_sum_path(root->left,maxi)); //return 0 if the sum is negative
    int right_sum = max(0,maximum_sum_path(root->right,maxi)); //same as above
    maxi = max(maxi, left_sum+right_sum+root->data);
    return (root->data) + max(left_sum, right_sum);
}
vector<vector<int>> zig_zag_traversal(BinaryTree* root){
    vector<vector<int>> res;
    if(root==nullptr){
        return res;
    }

    queue<BinaryTree*> q;
    q.push(root);
    bool flag = true; 

    while(!q.empty()){
        int size = q.size();
        vector<int> r(size);

        

        for (int i=0;i<size;i++){
            BinaryTree* node = q.front();
            q.pop();
            int index = flag ? i : size-1-i;

            r[index]=node->data;
            if(node->left){
                q.push(node->left);
            }
             if(node->right){
                q.push(node->right);
            }
        }
        flag = !flag;
        res.push_back(r);
    }
    return res;
}

int main(){

    BinaryTree* root=new BinaryTree(1);
    root->left = new BinaryTree(2);
    root->right = new BinaryTree(3);
    root->left->left = new BinaryTree(4);
    root->left->right = new BinaryTree(5);
    vector<vector<int>> result = zig_zag_traversal(root);
    for (const auto& level : result) {
    for (int val : level) {
        cout << val << " ";
    }
}

    // level_order_traversal(root);
    // inorder_traversal(root);
    // post_order_traversal(root);
    // pre_order_traversal(root);

    // cout<<"height of the given binary tree is: "<<height(root)-1;

    // vector<int> res=spiral_traversal(root);
    // for(auto i:res){
    //     cout<<i<<" ";
    // }

//     vector<int> ans=verticalTraversal(root);
//     for(auto i:ans){
//         cout<<i<<" ";
//     }
//    return 0;

    root->left = new BinaryTree(2);
    root->right = new BinaryTree(10);
    root->left->left = new BinaryTree(20);
    root->left->right = new BinaryTree(1);
    root->right->right = new BinaryTree(-25);
    root->right->right->left = new BinaryTree(3);
    root->right->right->right = new BinaryTree(4);

    // int maxi =0;
    // cout<<diameter_2(root,maxi);

    // BinaryTree* root1 = new BinaryTree(1);    
    // root1->left = new BinaryTree(2);   
    // root1->right = new BinaryTree(3); 
    // root1->left->left = new BinaryTree(4); 

    // BinaryTree* root2 = new BinaryTree(1);    
    // root2->left = new BinaryTree(2);   
    // root2->right = new BinaryTree(3); 
    // root2->left->left = new BinaryTree(41); 

    //  if (identical(root1, root2))
    //     cout << "Yes\n";
    // else
    //     cout << "No\n";


}

File: code/Tree.py
from collections import deque
class Node:
    def __init__(self,value):
        self.value=value
        self.left=None
        self.right=None
        

def height(root):
    
    if(root==None):
        return 0
    
    lheight=height(root.left)
    rheight=height(root.right)
    
    if(lheight<rheight):
        return rheight+1
    else:
        return lheight+1

def level_order_traversing(root):
    h=height(root)
    
    for i in range(1,h+1):
        level_order(root,i)
        
def level_order(root,level):
    if root==None:
        return 
    if level==1:
        return root.value
    elif(level>1):
        level_order(root.left,level-1)
        level_order(root.right,level-1)
    
    
def similar(root1,root2):
    h1=height(root1)
    h2=height(root2)
    
    if(h1==h2):
        for i in range(1,h1+1):
            if level_order(root1,h1)!=level_order(root2,h2):
                return False
    return True

def preorder_traversing(root):
    h=height(root)
    
    for i in range(1,h+1):
        printPreorder(root,i)
        
def printPreorder(root):
 
    if root:
 
        # First print the data of node
        print(root.value, end=" "),
 
        # Then recur on left child
        printPreorder(root.left)
 
        # Finally recur on right child
        printPreorder(root.right)
        
              
def Identical(root1,root2):
    if (root1==None) & (root2==None):
        return True
 
    if (root1!=None) & (root2!=None):
 
        return ((root1.value!=root2.value) and
        Identical(root1.left,root2.left)and
        Identical(root1.right,root2.right))
        
    else:
        return False
            
def morris_traversal(root):
    current = root   
    
    while(current):
        if current.left is None:
            print(current.value,end=" ")
            current = current.right
        else:
            pre=current.left
            
            while(pre.right and pre.right!=current):
                pre=pre.right
                
            if pre.right is None:
                pre.right=current
                current = current.left
            else:
                pre.right=None
                print(current.value,end=" ")
                current = current.right  
    
def level_order_traversal(root):
    h=height(root) 
    even=False
    odd=False
    for i in range(1,h+1):
        if(i%2==0):
            even=current_level_order_traversal(root,i)
            print(even)
        if(i%2!=0):
            odd=current_level_order_traversal(root,i)
    # return even,odd
            
        
                 
def current_level_order_traversal(root,level):
    if root is None:
        return 
    if level%2==0:
        return root.value
    
    if level%2!=0:
        return root.value
    else:
        current_level_order_traversal(root.left,level-1)
        current_level_order_traversal(root.right,level-1)
        
def queue_level_order(root):
    # if root is None:
    #         return
    queue=deque()
    queue.append(root)
    while(len(queue)>0):
        print(queue[0].value,end=" ")
        node=queue.popleft()
        if node.left is not None:
            queue.append(node.left)
        if node.right is not None:
            queue.append(node.right)
         
        
def check_even_odd(root):
    queue=deque()
    queue.append(root)
    level=0
    while(len(queue)>0):
        
        
        for i in range(len(queue)):
            node=queue[0]
            queue.popleft()
            
            if(level%2==0):#even
                if(node.value%2==0):
                    return False
            elif level%2!=0:
                if(node.value%2==1):
                    return False
            if node.left is not None:
                queue.append(node.left)
            if node.right is not None:
                queue.append(node.right)
        level+=1
    return True

def successor(root,val):
    succ=None
    while(root!=None):
        if(root.value<=val):
            root=root.right
        elif(root.value>val):
            succ=root
            root=root.left
    if succ is not None:
        return succ.value
    else:
        return None

def predecessor(root,val):
    pred=None
    while(root!=None):
        if(root.value>=val):
            root=root.left
        elif(root.value<val):
            pred=root
            root=root.right
    return pred.value
          

def a(root):
    queue=deque()
    queue.append(root)
    while(len(queue)>0):
        
        print(queue[0].value,end=" ")
        node=queue.popleft()

        if(node.left is not None):
            queue.append(node.left)
        if(node.right is not None):
            queue.append(node.right)  
root1=Node(5)
root1.left=Node(3)
root1.right=Node(7)
root1.left.left=Node(2)
root1.left.right=Node(4)
root1.right.right=Node(9)
root1.right.right.left=Node(8)
root1.right.right.right=Node(10)

root2=Node(1)
root2.left=Node(2)
root2.right=Node(3)
root2.right.left=Node(4)
root2.right.right=Node(5)

# i=10
# print(predecessor(root1,i))
# print(successor(root1,i))

# level_order_traversal(root1)
# if (check_even_odd(root1)):
#         print("YES")
# else: 
#     print("NO")

a(root2)